docs/architecture/adr-019-protobuf-state-encoding.md

```diff
@@ -110,162 +111,178 @@ type Codec interface {
-    cosmos_sdk.x.auth.vesting.v1.PeriodicVestingAccount   periodic_vesting_account   = 4;
-    cosmos_sdk.x.supply.v1.ModuleAccount                  module_account             = 5;
-  }
-
-  // ...
-}
-```
+### Usage of `Any` to encode interfaces
+
+In general, module-level .proto files should define messages which encode interfaces
+using [`google.protobuf.Any`](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/any.proto).
+After [extension discussion](https://github.com/cosmos/cosmos-sdk/issues/6030),
+this was chosen as the preferred alternative to application-level `oneof`s
+as in our original protobuf design. The arguments in favor of `Any` can be
+summarized as follows:
+
+* `Any` provides a simpler, more consistent client UX for dealing with
+interfaces than app-level `oneof`s that will need to be coordinated more
+carefully across applications. Creating a generic transaction
+signing library using `oneof`s may be cumbersome and critical logic may need
+to be reimplemented for each chain
+* `Any` provides more resistance against human error than `oneof`
+* `Any` is generally simpler to implement for both modules and apps
+
+The main counter-argument to using `Any` centers around its additional space
+and possibly performance overhead. The space overhead could be dealt with using
+compression at the persistence layer in the future and the performance impact
+is likely to be small. Thus, not using `Any` is seem as a pre-mature optimization,
+with user experience as the higher order concern.
+
+Note, that given the SDK's decision to adopt the `Codec` interfaces described
+above, app's can still choose to use `oneof` to encode state and transactions
+but it is not the recommended approach.
+
+### Safe usage of `Any`
+
+By default, the [gogo protobuf implementation of `Any`](https://godoc.org/github.com/gogo/protobuf/types)
+uses global type registration to decode values packed in `Any` into concrete
+go types. This introduces a vulnerability where any malicious module
+in the dependency tree could registry a type with the global protobuf registry
+and cause it to be loaded and unmarshaled by a transaction that referenced
+it in the `type_url` field.
+
+To prevent this, we introduce a type registration mechanism for decoding `Any`
+values into concrete types through the `InterfaceRegistry` interface which
+bears some similarity to type registration with Amino:
 
 ```go
-// app/codec/codec.go
-
-type Codec struct {
-  codec.Marshaler
-
-
-  amino *codec.Codec
-}
-
-func NewAppCodec(amino *codec.Codec) *Codec {
-  return &Codec{Marshaler: codec.NewHybridCodec(amino), amino: amino}
+type InterfaceRegistry interface {
+    // RegisterInterface associates protoName as the public name for the
+    // interface passed in as iface
+    // Ex:
+    //   ctx.RegisterInterface("cosmos_sdk.Msg", (*sdk.Msg)(nil))
+    RegisterInterface(protoName string, iface interface{})
```

Comment:
> q: Why do you need to register a `name` to an interface?

This is so that we can communicate which types satisfy which interface to clients. It is a usability thing. Since interfaces do not have a proto message name, we are giving them one here rather than making the client UX dependent on the golang name.

> pp: this feels over complicated and may create some confusion for example which module/ code is responsible to register an interface. Would there be sanity checks if all interfaces have an implementation for example, failures on duplicates or multiple names, ... must register interface be called before register implementation? It would be easier to drop this functionality.

There should be failures on duplicates but it would not need to be called before registering implementations.

Yes it would be easier to drop this and dropping it would also diminish client UX IMHO.

Reply:
Not depending on the Golang name is a useful feature when integrating message protos across an application with many pieces outside of the blockchain written in other languages.  It simplifies the ability to release proto bindings from a single shared reference project.